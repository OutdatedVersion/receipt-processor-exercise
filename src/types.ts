import { z } from 'zod';

export type ProcessedReceiptLedgerEntry = {
  ruleName: string;
  ruleVersion: number;
  pointsAwarded: number;
};

export type ProcessedReceipt = {
  id: string;
  pointsAwarded: number;
  ledger: ProcessedReceiptLedgerEntry[];
  receipt: Receipt;
};

export type Storage = {
  receipts: Map<string, ProcessedReceipt>;
};

const money = z.coerce.number().positive().finite();
export const receiptSchema = z.object({
  // Deviation: I added `&` to allow provided examples
  // ref https://github.com/fetch-rewards/receipt-processor-challenge/pull/11#issuecomment-1927953077
  retailer: z.string().regex(/^[\w\s-&]+$/),
  // TODO: check against yaml spec
  purchaseDate: z.string().regex(/\d{4}-\d{2}-\d{2}/),
  // TODO: check against yaml spec
  purchaseTime: z.string().regex(/\d{2}:\d{2}/),
  // TODO: does not strictly match spec
  total: money,
  items: z.array(
    z.object({
      shortDescription: z.string().regex(/^[\w\s-]+$/),
      price: money,
    }),
  ),
});
export type Receipt = z.infer<typeof receiptSchema>;
