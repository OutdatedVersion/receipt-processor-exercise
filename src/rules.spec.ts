import { describe, expect, it } from 'vitest';
import {
  alphanumericRule,
  applyRules,
  DefaultRules,
  everyTwoItemsRule,
  oddPurchaseDateRule,
  roundDollarAmountTotalRule,
  totalMultipleOfPoint25Rule,
} from './rules';
import { receiptSchema } from './types';

describe('rules overview', () => {
  it('matches the target example', () => {
    const receipt = receiptSchema.parse({
      retailer: 'Target',
      purchaseDate: '2022-01-01',
      purchaseTime: '13:01',
      items: [
        {
          shortDescription: 'Mountain Dew 12PK',
          price: '6.49',
        },
        {
          shortDescription: 'Emils Cheese Pizza',
          price: '12.25',
        },
        {
          shortDescription: 'Knorr Creamy Chicken',
          price: '1.26',
        },
        {
          shortDescription: 'Doritos Nacho Cheese',
          price: '3.35',
        },
        {
          shortDescription: '   Klarbrunn 12-PK 12 FL OZ  ',
          price: '12.00',
        },
      ],
      total: '35.35',
    });

    const result = applyRules(receipt, DefaultRules);

    expect(result.pointsAwarded).toBe(28);
    expect(result.ledger).toMatchSnapshot();
  });

  it('matches the M&M example', () => {
    const receipt = receiptSchema.parse({
      retailer: 'M&M Corner Market',
      purchaseDate: '2022-03-20',
      purchaseTime: '14:33',
      items: [
        {
          shortDescription: 'Gatorade',
          price: '2.25',
        },
        {
          shortDescription: 'Gatorade',
          price: '2.25',
        },
        {
          shortDescription: 'Gatorade',
          price: '2.25',
        },
        {
          shortDescription: 'Gatorade',
          price: '2.25',
        },
      ],
      total: '9.00',
    });

    const result = applyRules(receipt, DefaultRules);

    expect(result.pointsAwarded).toBe(109);
    expect(result.ledger).toMatchSnapshot();
  });
});

describe('individual rules', () => {
  describe('alphanumeric', () => {
    it('works', () => {
      expect(
        // @ts-expect-error
        alphanumericRule.apply({
          retailer: ' h1&%$ ',
        }),
      ).toBe(2);
    });
  });

  describe('multiple of 0.25', () => {
    it.each([0, -50])('does not explode for %s', (total) => {
      // @ts-expect-error
      totalMultipleOfPoint25Rule.apply({
        total,
      });
    });

    it.each([3.51, 33.33])('works for negative cases: %s', (total) => {
      expect(
        // @ts-expect-error
        totalMultipleOfPoint25Rule.apply({
          total,
        }),
      ).toBe(0);
    });

    it.each([2, 40, 600, 4.5])('works for positive cases: %s', (total) => {
      expect(
        // @ts-expect-error
        totalMultipleOfPoint25Rule.apply({
          total,
        }),
      ).toBe(25);
    });
  });

  describe('round dollar total', () => {
    it.each([0, -520])('does not explode for %s', (total) => {
      // @ts-expect-error
      roundDollarAmountTotalRule.apply({
        total,
      });
    });

    it.each([4.2, 3.51, 33.33])('works for negative cases: %s', (total) => {
      expect(
        // @ts-expect-error
        roundDollarAmountTotalRule.apply({
          total,
        }),
      ).toBe(0);
    });

    it.each([50, 259372, 1, 2])('works for positive cases: %s', (total) => {
      expect(
        // @ts-expect-error
        roundDollarAmountTotalRule.apply({
          total,
        }),
      ).toBe(50);
    });
  });

  describe('every two items', () => {
    it('is ok with empty items', () => {
      expect(
        // @ts-expect-error
        everyTwoItemsRule.apply({ items: [] }),
      ).toBe(0);
    });

    it('works', () => {
      expect(
        // @ts-expect-error
        everyTwoItemsRule.apply({
          items: [{ price: 1, shortDescription: 'idk' }],
        }),
      ).toBe(0);

      expect(
        // @ts-expect-error
        everyTwoItemsRule.apply({
          items: [
            { price: 1, shortDescription: 'idk' },
            { price: 2, shortDescription: 'yeah' },
          ],
        }),
      ).toBe(5);

      expect(
        // @ts-expect-error
        everyTwoItemsRule.apply({
          items: [
            { price: 1, shortDescription: 'idk' },
            { price: 2, shortDescription: 'yeah' },
            { price: 7, shortDescription: 'yes' },
            { price: 24, shortDescription: 'nope' },
          ],
        }),
      ).toBe(10);
    });
  });

  describe('odd purchase date', () => {
    it('does not explode on junk', () => {
      expect(
        // @ts-expect-error
        oddPurchaseDateRule.apply({ purchaseDate: 'um' }),
      ).toBe(0);
      expect(
        // @ts-expect-error
        oddPurchaseDateRule.apply({ purchaseDate: '10-10-1998' }),
      ).toBe(0);
    });

    it('works', () => {
      expect(
        // @ts-expect-error
        oddPurchaseDateRule.apply({ purchaseDate: '2000-01-01' }),
      ).toBe(6);
      expect(
        // @ts-expect-error
        oddPurchaseDateRule.apply({ purchaseDate: '2000-01-02' }),
      ).toBe(0);
    });
  });
});
