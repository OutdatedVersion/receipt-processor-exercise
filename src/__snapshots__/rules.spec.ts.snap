// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rules overview > matches the M&M example 1`] = `
[
  {
    "pointsAwarded": 14,
    "ruleName": "Point per alphanumeric character in retailer name",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 25,
    "ruleName": "Lump points if total is a multiple of 0.25",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 50,
    "ruleName": "Lump points if total is round dollar amount (no cents)",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 10,
    "ruleName": "Points for every 2 items",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 10,
    "ruleName": "Lump points if purchase time is between 2:00pm and 4:00pm",
    "ruleVersion": 1,
  },
]
`;

exports[`rules overview > matches the target example 1`] = `
[
  {
    "pointsAwarded": 6,
    "ruleName": "Point per alphanumeric character in retailer name",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 10,
    "ruleName": "Points for every 2 items",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 6,
    "ruleName": "Lump points if purchase date's day-of-month is odd number",
    "ruleVersion": 1,
  },
  {
    "pointsAwarded": 6,
    "ruleName": "Points for every item with a description whose length is a multiple of 3",
    "ruleVersion": 1,
  },
]
`;
